### ISSUE TITLE
Do not generate "readOnly" for get-only lists in schema

### ISSUE THREAD

**Issue by Dennis-Petrov** [domaindrivendev/Swashbuckle.AspNetCore#1536]
Greetings!
There is well-known language pattern for lazy-initialized get-only collections, which avoids paranoid null-checking: 

```
public class Order
{
    private IList<OrderItem> items;

    public IList<OrderItem> Items => items ?? (items = new List<OrderItem>());

    // ...
}

// No need to check Items for null
order.Items.Where(_ => _.Price > 100);

```
While `Items` is get-only property from the point of C#, it could be populated, serialized to JSON, posted to the server, deserialized, and then processed at server side.

At the moment properties like this are read-only in schema, and they are ignored in generated docs:
```
          "Items": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/OrderItem"
            },
            "nullable": true,
            "readOnly": true  
```
But logically they **are not** read-only.
IMO, such properties must not be defined as read-only in JSON schema by default.

P.S. I know, that I can write `ISchemaFilter` like this, but this issue is about default behavior:

```
    internal sealed class ReadonlyCollectionsFilter : ISchemaFilter
    {
        private static readonly Type GerericIList = typeof(IList<>);

        public void Apply(OpenApiSchema schema, SchemaFilterContext context)
        {
            if (schema.Properties == null || schema.Properties.Count == 0)
            {
                return;
            }

            foreach (var schemaProperty in schema.Properties.Where(_ => _.Value.ReadOnly))
            {
                var property = context.Type.GetProperty(schemaProperty.Key, BindingFlags.Public | BindingFlags.Instance);

                if (property.PropertyType.IsGenericType && property.PropertyType.GetGenericTypeDefinition() == GerericIList)
                {
                    schemaProperty.Value.ReadOnly = false;
                }
            }
        }
    }
```
I'm using Swashbuckle.AspNetCore 5.0.0.

**Comment by domaindrivendev** [comment]
Actually this is a little more nuanced than you think. Firstly, it's worth noting that the key here is "serializer behavior" and not just the property definition in C#. Because it's the serializer behavior that ultimately dictates what can and can't be provided to your API in a JSON payload.

Swashbuckle currently supports two serializer implementations - `Newtonsoft` and `System.Text.Json (STJ)`. While you're correct in saying that Newtonsoft WILL de-serialize a JSON array into a _readonly_ property of type `IList/List` , the same is NOT true for the STJ serializer, which is now the default serializer for ASP.NET Core. So, setting the `readOnly` flag here may or may not be accurate dependending on the serializer that's in use.

In the case of Newtonsoft, I agree that the flag _should_ not be set for those properties, assuming the lazy instantiation logic is in place. However, this is a non-trivial change for Swashbuckle. Detecting the type is easy but the lazy instantiation not so much. Swashbuckle would have to actually instantiate the DTO, something I'm loathe to do because it's effectively executing application code and this seems overly intrusive.

**Comment by Dennis-Petrov** [comment]
Hi.

> but the lazy instantiation not so much

I agree with you, that implementing universal solution will be rather complex.
But mentioned pattern is widely used.

That's why I suggest you to add an option, which will generate `"readOnly": false` when:
- serializer is `Newtonsoft.Json`
- property has not public setter (or has not setter at all)
- property type implements `IList` or `IList<T>`
- property type is not an array.

This will cover 99% of cases (I know about `IList.IsReadOnly`, this will be rest 1%).

**Comment by domaindrivendev** [comment]
I understand the pattern is used widely in general but how wide spread is it specifically for API models that are being deserialized from a JSON payload vs just including a setter?

My problem with what you've suggested is that it doesn't reflect actual de-serialization behavior. All of the criteria you listed above could be true but without the lazy instantiation logic, the Newtonsfto serializer will not populate the list property.

There may be an argument that the use of `IList` vs `IEnumerable` (or an array) is a strong indication that the developer is using this pattern, but I'm not convinced it's a 99% likelyhood or anything near that.

In your particular use-case, may I ask why you are so keen to avoid the public setter in this case? To be clear, I'm not a fan of public setters, particularly in classes where I'm trying to encapsulate business logic but for API models, where the classes sole purpose is to capture the incoming JSON payload, I'm generally less strict on this principle.

**Comment by Dennis-Petrov** [comment]
> but without the lazy instantiation logic, the Newtonsoft serializer will not populate the list property

AFAIK, during de-serialization NS.Json knows nothing about laziness. In fact, the list may not be *lazy*. It could be initialized in constructor. NS.Json just inspects property metadata: if property is get-only, and this is `IList`, and it is not read-only, NS.Json *tries* to fill it from JSON array.

> may I ask why you are so keen to avoid the public setter in this case?

It's code design issue.

Imagine a DTO with collection of orders, where each order contains several items, where each item contains several taxes, 

And then imagine LINQ query, which gets all used tax rates, :)
In case of get-only collections one don't need to check collection properties for nulls. The code is much more readable, and less chances to get NRE.

**Comment by domaindrivendev** [comment]
> AFAIK, during de-serialization NS.Json knows nothing about laziness. In fact, the list may not be lazy. It could be initialized in constructor. NS.Json just inspects property metadata: if property is get-only, and this is IList, and it is not read-only, NS.Json tries to fill it from JSON array.

Not laziness per se but whether or not the property has been initialized with a value at the time of deserialization. So yes, this could be in a constructor as well. But the point is the same - if this hasn't been done (i.e. if the value is NULL), then the serializer WILL NOT populate the property. Try it for yourself if you don't believe me

**Comment by domaindrivendev** [comment]
If it helps, I could add a custom property to the `Annotations` library that allows you to explicitly assign the `readOnly` and `writeOnly` flags. For example:

```csharp
[SwaggerProperty(ReadOnly = false)]
public IList<OrderItem> Items => items ?? (items = new List<OrderItem>());
```

**Comment by Dennis-Petrov** [comment]
> if this hasn't been done (i.e. if the value is NULL), then the serializer WILL NOT populate the property

Yes, this is true.

> I could add a custom attribute

Great, This will solve the problem.

**Comment by domaindrivendev** [comment]
A solution for this is now merged to master and queued up for the `v5.3.0` release (due next week). There's now a `SwaggerSchemaAttribute` that can be applied to classes OR properties. When you apply it to a property you can explicitly indicate if the property should be described as `readOnly` or `writeOnly`:

```csharp
public class Order
{
    private IList<OrderItem> items;

    [SwaggerSchema(ReadOnly = false)]
    public IList<OrderItem> Items => items ?? (items = new List<OrderItem>());
}
```

To use this you'll have to have the `Annotations` package installed and explicitly opted-in as described [here](https://github.com/domaindrivendev/Swashbuckle.AspNetCore#swashbuckleaspnetcoreannotations)
